os/kernel/debug/mem_leak_checker.c

```diff
@@ -0,0 +1,327 @@
+					continue;
+				}
+				if (search_hash(*(unsigned long volatile *)leak_chk - (unsigned long)SIZEOF_MM_ALLOCNODE)) {
+					(*leak_cnt)--;
+				}
+			}
+		}
+	}
+}
+
+static void ram_check(char *bin_name, int *leak_cnt, uint32_t *bin_text_addr)
+{
+#ifdef CONFIG_APP_BINARY_SEPARATION
+	bin_addr_info_t *info;
+	int bin_idx;
+
+	info = (bin_addr_info_t *)get_bin_addr_list();
+	for (bin_idx = 0; bin_idx <= CONFIG_NUM_APPS; bin_idx++) {
+		if (strncmp(BIN_NAME(bin_idx), bin_name, strlen(bin_name)) == 0) {
+			/* For calculating the owner from binary, save the text section address. */
+			*bin_text_addr = info[bin_idx].text_addr;
+		}
+	}
+#endif
+	/* Visit the RAM region, except heap region.
+	 * So check from RAM start to heap start, and from heap end to RAM end.
+	 */
+	search_addr(RAM_START_ADDR, HEAP_START_ADDR, leak_cnt);
+	search_addr(HEAP_END_ADDR, RAM_END_ADDR, leak_cnt);
+
+	/* Visit heap region */
+	heap_check(leak_cnt);
+}
+
+static void print_info(char *bin_name, int leak_cnt, int broken_cnt, uint32_t bin_text_addr)
+{
+	volatile struct mm_allocnode_s *node;
+	uint32_t owner_addr;	
+
+#ifdef CONFIG_APP_BINARY_SEPARATION
+	elf_show_all_bin_section_addr();
+#endif
+
+
+	if (leak_cnt > 0 || broken_cnt > 0) {
+		lldbg("Type   |    Addr   |   Size   |  Owner  | PID \n");
+		lldbg("---------------------------------------------------\n");
+
+		mm_takesemaphore((struct mm_heap_s *)leak_checker.heap);
+
+		for (node = (struct mm_allocnode_s *)HEAP_START_ADDR; node < (struct mm_allocnode_s *)HEAP_END_ADDR; node = (struct mm_allocnode_s *)((char *)node + node->size)) {
+			if (node->reserved == MEM_LEAK) {
+				/* alloc_call_addr can be from kernel, app or common binary.
+				 * based on the text addresses printed, user needs to check the
+				 * corresponding binaries accordingly 
+				 */
+				owner_addr = node->alloc_call_addr;
+				pid_t pid = node->pid;
+				if (pid < 0) {
+					/* For stack allocated node, pid is negative value.
+					 * To use the pid, change it to original positive value.
+					 */
+					pid = (-1) * pid;
+				}
+				lldbg("LEAK   |  %p  | %8d | %p | %d\n", (void *)((char *)node + SIZEOF_MM_ALLOCNODE), node->size - SIZEOF_MM_ALLOCNODE, owner_addr, pid);
+			} else if (node->reserved == MEM_BROKEN) {
+				lldbg("BROKEN | %p\n", node);
+			}
+		}
+
+		mm_givesemaphore((struct mm_heap_s *)leak_checker.heap);
+
+		lldbg("*** %d LEAKS, %d BROKENS.\n", leak_cnt, broken_cnt);
+	} else {
+		lldbg("*** NO MEMORY LEAK.\n");
+	}
+}
+
+static void init_mem_leak_checker(int checker_pid, char *bin_name)
+{
+	struct tcb_s *checker_tcb = sched_gettcb(checker_pid);
+	struct mm_heap_s *heap;
+
+	/* stack information is for leak checker task. */
+	leak_checker.stack_top = checker_tcb->adj_stack_ptr;
+	leak_checker.stack_bottom = leak_checker.stack_top - checker_tcb->adj_stack_size;
+#ifdef CONFIG_APP_BINARY_SEPARATION
+	if (strncmp(bin_name, "kernel", strlen("kernel")) == 0) {
```

Comment:
How about comparing with `strlen(kernel) + 1`?
This can check much more secure.

Reply:
Please check Jeongchan's comment above.